src/Qir/Runtime/stdlib/src/callables.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#![deny(clippy::all, clippy::pedantic)]
+
+use crate::{
+    arrays::{
+        QirArray, __quantum__rt__array_concatenate, __quantum__rt__array_update_reference_count,
+    },
+    tuples::{__quantum__rt__tuple_copy, __quantum__rt__tuple_update_reference_count},
+    update_counts,
+};
+use std::{cell::RefCell, rc::Rc, usize};
+
+#[derive(Clone)]
+pub struct Callable {
+    func_table: *mut *mut u8,
+    mem_table: *mut *mut u8,
+    cap_tuple: *mut u8,
+    is_adj: RefCell<bool>,
+    is_ctl: RefCell<u32>,
+}
+
+#[no_mangle]
+pub extern "C" fn __quantum__rt__callable_create(
+    func_table: *mut *mut u8,
+    mem_table: *mut *mut u8,
+    cap_tuple: *mut u8,
+) -> *const Callable {
+    Rc::into_raw(Rc::new(Callable {
+        func_table,
+        mem_table,
+        cap_tuple,
+        is_adj: RefCell::new(false),
+        is_ctl: RefCell::new(0),
+    }))
+}
+
+/// # Safety
+///
+/// This function should only be called with a callable created by `__quantum__rt__callable_*` functions.
+#[no_mangle]
+#[allow(clippy::cast_ptr_alignment)]
+pub unsafe extern "C" fn __quantum__rt__callable_invoke(
+    callable: *const Callable,
+    args_tup: *mut u8,
+    res_tup: *mut u8,
+) {
+    let call = Rc::from_raw(callable);
+    let index = (if *call.is_adj.borrow() { 1 } else { 0 })
+        + (if *call.is_ctl.borrow() > 0 { 2 } else { 0 });
+
+    // Collect any nested controls into a single control list.
+    let mut args_copy: *mut *const Vec<u8> = std::ptr::null_mut();
+    if !args_tup.is_null() {
+        // Copy the tuple so we can potentially edit it.
+        args_copy = __quantum__rt__tuple_copy(args_tup.cast::<*const Vec<u8>>(), true);
+
+        if *call.is_ctl.borrow() > 0 {
```

Comment:
Ah, does `is_ctl` represent the number of controls? I may suggest a more descriptive name like `n_control_qubits`, then?

Reply:
I went with `ctls_count`, good call on clarifying the name!